src/js/libcs/Render2D.js

```diff
@@ -373,30 +373,22 @@ Render2D.arrowShapes = {
     },
 };
 
-Render2D.onSegment = function(startpos, endpos, pt) {
-    var det = List.det(List.realMatrix([
-        [startpos.x, startpos.y, 1],
-        [endpos.x, endpos.y, 1],
-        [pt.x, pt.y, 1]
-    ]));
-    var onLine = CSNumber._helper.isAlmostEqual(det, CSNumber.zero, CSNumber.epsbig);
-    var a = List.realVector([startpos.x, startpos.y, 1]);
-    var b = List.realVector([endpos.x - startpos.x, endpos.y - startpos.y, 0]);
-    var c = List.realVector([pt.x, pt.y, 1]);
-    var d = List.realVector([endpos.x, endpos.y, 1]);
-    var cr = List.crossratio3(a, b, c, d, List.ii).value.real;
-    return onLine && cr >= 0 && cr <= 1;
-};
-
 Render2D.clipSegment = function(pt1, pt2) {
-    var clipPoints = Render2D.clipLineCore(pt1.y - pt2.y, pt2.x - pt1.x, pt1.x * pt2.y - pt2.x * pt1.y);
+    var dx = pt2.x - pt1.x;
+    var dy = pt2.y - pt1.y;
+    var clipPoints = Render2D.clipLineCore(-dy, dx, pt1.x * pt2.y - pt2.x * pt1.y);
     if (clipPoints.length !== 2) return [];
-    var res = [];
-    if (Render2D.onSegment(clipPoints[0], clipPoints[1], pt1)) res.push(pt1);
-    if (Render2D.onSegment(clipPoints[0], clipPoints[1], pt2)) res.push(pt2);
-    if (Render2D.onSegment(pt1, pt2, clipPoints[0])) res.push(clipPoints[0]);
-    if (Render2D.onSegment(pt1, pt2, clipPoints[1])) res.push(clipPoints[1]);
-    return res;
+    var q1 = clipPoints[0];
+    var q2 = clipPoints[1];
+    var factor = 1 / (dx * dx + dy * dy);
```

Comment:
OK, in that case we call `Render2D.clipLineCore(0, 0, 0)`, which returns `[]`, so that we never reach line 383 due to the return in line 380. So everything is fine.